File: cxx11-crow/infix.cxx
#include <cctype>
#include <list>
#include <stack>
#include <string>
#include <iostream>

#include "rpn.hxx"

using namespace std;

namespace infixcalc {
  int precedence(char op) {
    switch (op) {
    case '+':
    case '-':
      return 1;
    case '*':
    case '/':
      return 2;
    default: // including '(', which is the most important one
      return 0;
    }
  }

  list<string> convertInfix(list<string> tokens) {
    stack<string> operators;
    list<string> result;

    for (auto token: tokens) {
      if (token.size() != 1 || isdigit(token[0])) {
        // we assume it's a number
        result.push_back(token);
      } else {
        switch (token[0]) {
        default:
          throw;
        case '(':
          operators.push(token);
          break;
        case ')':
          while (operators.top() != "(") {
            if (operators.size() == 0) {
              throw;
            }
            result.push_back(operators.top());
            operators.pop();
          }
          operators.pop();
          break;
        case '+':
        case '-':
        case '*':
        case '/':
          int tokenPrecedence = precedence(token[0]);
          int stackPrecedence;
          do {
            if (operators.size() == 0) {
              operators.push(token);
              break;
            }
            stackPrecedence = precedence(operators.top()[0]);

            if (tokenPrecedence <= stackPrecedence) {
              result.push_back(operators.top());
              operators.pop();
            }

            if (tokenPrecedence >= stackPrecedence) {
              operators.push(token);
            }
          } while (tokenPrecedence < stackPrecedence);
          break;
        }
      }
    }

    while (operators.size() > 0) {
      result.push_back(operators.top());
      operators.pop();
    }

    return result;
  }

  double evaluateInfix(list<string> tokens) {
    return rpncalc::evaluateRpn(convertInfix(tokens));
  }
};

int main() {
  list<string> testexpr {"4", "*", "(", "2", "+", "3", ")"};

  cout << infixcalc::evaluateInfix(testexpr) << endl;
  auto result = infixcalc::convertInfix(testexpr);
  for (auto token : result) {
    cout << token << ' ';
  }
  cout << endl;
}

File: golang/src/github.com/DanielBrookRoberge/rpn/infix.go
package rpn

import "github.com/alediaferia/stackgo"
import "strconv"

var precedence = map[string]int{
	"^": 3,
	"*": 2,
	"/": 2,
	"-": 1,
	"+": 1,
	"(": 0,
}

func ConvertInfix(tokens chan string) chan string {
	c := make(chan string, 5)

	go func() {
		stack := stackgo.NewStack()

		for token := range tokens {
			if _, err := strconv.ParseFloat(token, 64); err == nil {
				c <- token
			} else if token == "(" {
				stack.Push(token)
			} else if token == ")" {
				// This will panic if the stack underflows
				for stack.Top() != "(" {
					c <- stack.Pop().(string)
				}
				stack.Pop()
			} else {
				tokenPrecedence, ok := precedence[token]
				if !ok {
					panic("Invalid token")
				}
				for {
					if stack.Top() == nil {
						stack.Push(token)
						break
					}
					stackPrecedence := precedence[stack.Top().(string)]

					if tokenPrecedence <= stackPrecedence {
						c <- stack.Pop().(string)
					}

					if tokenPrecedence >= stackPrecedence {
						stack.Push(token)
						break
					}
				}

			}
		}

		for stack.Top() != nil {
			c <- stack.Pop().(string)
		}
		close(c)
	}()

	return c
}
